File: clients/csharp/Model/BuildingType.cs
namespace SpbAiChamp.Model
{
    /// <summary>
    /// Building type
    /// </summary>
    public enum BuildingType
    {
        /// <summary>
        /// Quarry harvests stone
        /// </summary>
        Quarry = 0,
        /// <summary>
        /// Mines harvests ore
        /// </summary>
        Mines = 1,
        /// <summary>
        /// Career harvest sand
        /// </summary>
        Career = 2,
        /// <summary>
        /// Farm harvests organics
        /// </summary>
        Farm = 3,
        /// <summary>
        /// Foundry produces metal
        /// </summary>
        Foundry = 4,
        /// <summary>
        /// Furnace produces silicon
        /// </summary>
        Furnace = 5,
        /// <summary>
        /// Bioreactor produces plastic
        /// </summary>
        Bioreactor = 6,
        /// <summary>
        /// Chip factory produces chips
        /// </summary>
        ChipFactory = 7,
        /// <summary>
        /// Accumulator factory produces accumulators
        /// </summary>
        AccumulatorFactory = 8,
        /// <summary>
        /// Replicator produces new workers
        /// </summary>
        Replicator = 9,
    }

    public static class BuildingTypeHelper {
        /// <summary> Read BuildingType from reader </summary>
        public static BuildingType ReadFrom(System.IO.BinaryReader reader) {
            switch (reader.ReadInt32())
            {
                case 0:
                    return BuildingType.Quarry;
                case 1:
                    return BuildingType.Mines;
                case 2:
                    return BuildingType.Career;
                case 3:
                    return BuildingType.Farm;
                case 4:
                    return BuildingType.Foundry;
                case 5:
                    return BuildingType.Furnace;
                case 6:
                    return BuildingType.Bioreactor;
                case 7:
                    return BuildingType.ChipFactory;
                case 8:
                    return BuildingType.AccumulatorFactory;
                case 9:
                    return BuildingType.Replicator;
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }
    }
}

File: clients/scala/src/main/scala/spb_ai_champ/model/BuildingType.scala
package spb_ai_champ.model

import spb_ai_champ.util.StreamUtil

/**
 * Building type
 */
sealed abstract class BuildingType (val tag: Int) {
    /**
     * Write BuildingType to output stream
     */
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, tag)
    }
}

object BuildingType {
    /**
     * Quarry harvests stone
     */
    case object QUARRY extends BuildingType(0)
    /**
     * Mines harvests ore
     */
    case object MINES extends BuildingType(1)
    /**
     * Career harvest sand
     */
    case object CAREER extends BuildingType(2)
    /**
     * Farm harvests organics
     */
    case object FARM extends BuildingType(3)
    /**
     * Foundry produces metal
     */
    case object FOUNDRY extends BuildingType(4)
    /**
     * Furnace produces silicon
     */
    case object FURNACE extends BuildingType(5)
    /**
     * Bioreactor produces plastic
     */
    case object BIOREACTOR extends BuildingType(6)
    /**
     * Chip factory produces chips
     */
    case object CHIP_FACTORY extends BuildingType(7)
    /**
     * Accumulator factory produces accumulators
     */
    case object ACCUMULATOR_FACTORY extends BuildingType(8)
    /**
     * Replicator produces new workers
     */
    case object REPLICATOR extends BuildingType(9)

    /**
     * Read BuildingType from input stream
     */
    def readFrom(stream: java.io.InputStream): BuildingType =
        StreamUtil.readInt(stream) match {
            case QUARRY.tag => QUARRY
            case MINES.tag => MINES
            case CAREER.tag => CAREER
            case FARM.tag => FARM
            case FOUNDRY.tag => FOUNDRY
            case FURNACE.tag => FURNACE
            case BIOREACTOR.tag => BIOREACTOR
            case CHIP_FACTORY.tag => CHIP_FACTORY
            case ACCUMULATOR_FACTORY.tag => ACCUMULATOR_FACTORY
            case REPLICATOR.tag => REPLICATOR
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
}